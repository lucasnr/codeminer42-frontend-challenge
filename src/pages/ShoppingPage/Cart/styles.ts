import styled from 'styled-components';

export const Container = styled.section`
	background-color: rgba(0, 0, 0, 0.075);
	border-radius: 0.5rem;
	overflow: hidden;
`;

export const Title = styled.h2`
	background-color: ${({ theme }) => theme.colors.secondary};
	color: ${({ theme }) => theme.text.secondary};
	font-size: 1.5rem;
	font-weight: 600;
	padding: 0.5rem;
	text-align: center;
`;

export const Content = styled.div`
	padding: 1rem;
`;

export const Discount = styled.form`
	display: flex;

	input {
		border: 0.0625rem solid rgba(0, 0, 0, 0.5);
		border-radius: 0.25rem;
		font-size: 0.875rem;
		font-weight: 400;
		margin-right: 1rem;
		padding: 0.625rem 1rem;
		width: 100%;
	}

	button {
		background-color: ${({ theme }) => theme.colors.primary};
		border: none;
		border-radius: 0.5rem;
		color: ${({ theme }) => theme.text.secondary};
		font-size: 0.875rem;
		font-weight: 500;
		padding: 0.625rem 1rem;
		text-transform: uppercase;
	}
`;

export const Totals = styled.ul`
	li {
		align-items: center;
		border-top: 0.0625rem solid rgba(0, 0, 0, 0.125);
		display: flex;
		font-size: 1rem;
		justify-content: space-between;
		padding: 1rem;

		&:last-child {
			font-weight: 700;
		}
	}
`;
