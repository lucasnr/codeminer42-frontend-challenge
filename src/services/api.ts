import axios from 'axios';

import { IVoucher } from '~/store/modules/cart/types';

const api = axios.create({
	baseURL: 'https://shielded-wildwood-82973.herokuapp.com',
});

export interface IApiProduct {
	id: number;
	name: string;
	price: number;
	available: number;
}

export interface FetchProductsServerResponse {
	products: IApiProduct[];
}

export interface FetchVouchersServerResponse {
	vouchers: IVoucher[];
}

export const getProducts = () =>
	api.get<FetchProductsServerResponse>('products.json');

export const getVouchers = () =>
	api.get<FetchVouchersServerResponse>('vouchers.json');
