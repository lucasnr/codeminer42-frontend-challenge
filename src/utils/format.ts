const locale = 'pt-BR'; // 1.000,00
const moneyFormatter = new Intl.NumberFormat(locale, {
	minimumFractionDigits: 2,
});

export function formatMoney(number: number) {
	return moneyFormatter.format(number);
}
