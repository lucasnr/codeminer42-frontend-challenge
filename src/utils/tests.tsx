import React, { ReactElement } from 'react';
import { mount } from 'enzyme';
import { ThemeProvider } from 'styled-components';
import { Provider } from 'react-redux';
import { MockStoreEnhanced } from 'redux-mock-store';

import light from '~/styles/themes/light';

import { ApplicationState } from '~/store';

export const flushPromises = () => new Promise(setImmediate);

export const mountWithProviders = (
	children: ReactElement,
	store?: MockStoreEnhanced<ApplicationState>
) => {
	if (store)
		return mount(
			<Provider store={store}>
				<ThemeProvider theme={light}>{children}</ThemeProvider>
			</Provider>
		);
	else return mount(<ThemeProvider theme={light}>{children}</ThemeProvider>);
};
