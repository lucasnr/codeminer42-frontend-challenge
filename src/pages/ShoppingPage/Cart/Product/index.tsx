import React, { useMemo } from 'react';

import { Container, Content, Name, Row, Controls } from './styles';

import getImage from '~/utils/images';

import { IProduct } from '~/store/modules/cart/types';

const Product: React.FC<IProduct> = ({ id, name, quantity, price }) => {
	const image = useMemo(() => getImage(id), [id]);

	return (
		<Container>
			<img alt={name} src={image} />

			<Content>
				<Name>{name}</Name>
				<Row>
					<span>Quantity: {quantity}</span>
					<span>$ {price}</span>
				</Row>
			</Content>
			<Controls>
				<button>+</button>
				<button>-</button>
			</Controls>
		</Container>
	);
};

export default Product;
