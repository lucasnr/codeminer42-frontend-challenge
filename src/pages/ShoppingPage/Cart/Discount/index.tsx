import React, { useCallback, useState } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { toast } from 'react-toastify';

import { Container } from './styles';

import api from '~/services/api';

import { ApplicationState } from '~/store';
import { IVoucher } from '~/store/modules/cart/types';
import { addVoucher } from '~/store/modules/cart/actions';

const Discount: React.FC = () => {
	const [handlingSubmit, setHandlingSubmit] = useState(false);
	const [value, setValue] = useState('');

	const vouchers = useSelector(
		(state: ApplicationState) => state.cart.vouchers
	);
	const dispatch = useDispatch();
	const handleAddVoucher = useCallback(
		async (event: React.FormEvent) => {
			event.preventDefault();
			if (handlingSubmit) return;

			setHandlingSubmit(true);
			try {
				const { data } = await api.get('vouchers.json');
				const apiVouchers: IVoucher[] = data.vouchers;

				const voucher = apiVouchers.find((voucher) => voucher.code === value);
				if (!voucher) {
					toast.warning('No voucher was found with this code');
					return;
				}

				const exists = vouchers.find(({ id }) => voucher.id === id);
				if (!exists) {
					dispatch(addVoucher(voucher));
					toast.success('Voucher added');
				} else toast.warning('This voucher has already been added');
			} catch (error) {
				toast.error('Something went wrong. Please try again');
			} finally {
				setHandlingSubmit(false);
			}
		},
		[dispatch, handlingSubmit, value, vouchers]
	);

	return (
		<Container onSubmit={handleAddVoucher}>
			<input
				type="text"
				placeholder="Discount code"
				onChange={(event) => setValue(event.target.value)}
				value={value}
			/>
			<button type="submit">Apply</button>
		</Container>
	);
};

export default Discount;
