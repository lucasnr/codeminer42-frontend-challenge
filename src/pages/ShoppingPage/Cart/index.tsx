import React from 'react';
import { useSelector } from 'react-redux';

import { ApplicationState } from '~/store';

import Product from './Product';
import Discount from './Discount';
import { Container, Title, Content, Totals, Vouchers, Voucher } from './styles';

const Cart: React.FC = () => {
	const {
		subtotal,
		shipping,
		discount,
		total,
		products,
		vouchers,
	} = useSelector((state: ApplicationState) => state.cart);

	return (
		<Container>
			<Title>Shopping Cart</Title>
			<Content>
				<ul>
					{products.map((product) => (
						<Product key={product.id} {...product} />
					))}
				</ul>

				<Discount />
			</Content>

			<Totals>
				<li>
					<span>Subtotal</span>
					<span>$ {subtotal}</span>
				</li>
				<li>
					<span>Shipping</span>
					<span>$ {shipping}</span>
				</li>
				<Vouchers>
					<span>Discount</span>
					{vouchers.map((voucher) => (
						<Voucher key={voucher.id}>{voucher.code}</Voucher>
					))}
					<span>$ {discount}</span>
				</Vouchers>
				<li>
					<span>Total</span>
					<span>$ {total}</span>
				</li>
			</Totals>
		</Container>
	);
};

export default Cart;
