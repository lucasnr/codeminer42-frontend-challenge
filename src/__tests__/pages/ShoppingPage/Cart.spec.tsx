import React from 'react';
import { ReactWrapper } from 'enzyme';
import { act } from 'react-dom/test-utils';
import { AxiosResponse } from 'axios';
import configureStore, { MockStoreEnhanced } from 'redux-mock-store';

import Cart from '~/pages/ShoppingPage/Cart';
import Discount from '~/pages/ShoppingPage/Cart/Discount';

import { formatMoney } from '~/utils/format';
import { mountWithProviders, flushPromises } from '~/utils/tests';

import { Provider } from 'react-redux';
import realStore, { ApplicationState } from '~/store';

import * as api from '~/services/api';

const mockStore = configureStore<ApplicationState>([]);
const initialState: ApplicationState = {
	cart: {
		products: [
			{
				id: 5,
				name: 'Strawberry',
				price: 15,
				available: 9,
				quantity: 1,
			},
		],
		subtotal: 15,
		shipping: 30,
		discount: 0,
		vouchers: [],
		total: 45,
	},
};

describe('<Cart>', () => {
	let store: MockStoreEnhanced<ApplicationState>;
	let wrapper: ReactWrapper;

	beforeEach(async () => {
		store = mockStore(initialState);

		await act(async () => {
			wrapper = mountWithProviders(<Cart />, store);
		});
	});

	it('renders without fail and matches snapshot', async () => {
		expect(wrapper).toMatchSnapshot();
	});

	it('renders correct values from store', async () => {
		const expectedSubtotal = `$ ${formatMoney(15)}`;
		const expectedShipping = `$ ${formatMoney(30)}`;
		const expectedDiscount = `$ ${formatMoney(0)}`;
		const expectedTotal = `$ ${formatMoney(45)}`;

		expect(wrapper.find('[data-testid="subtotal"]').text()).toBe(
			expectedSubtotal
		);
		expect(wrapper.find('[data-testid="shipping"]').text()).toBe(
			expectedShipping
		);
		expect(wrapper.find('[data-testid="discount"]').text()).toBe(
			expectedDiscount
		);
		expect(wrapper.find('[data-testid="total"]').text()).toBe(expectedTotal);
	});

	it('adds voucher', async () => {
		const code = '#50OFF';

		const fakeResponse: AxiosResponse<api.FetchVouchersServerResponse> = {
			data: {
				vouchers: [
					{
						id: 2,
						code,
						amount: 50,
						type: 'percentual',
					},
				],
			},
			status: 200,
			statusText: 'Ok',
			headers: {
				'content-type': 'application/json',
			},
			config: {},
		};
		const apiSpy = jest
			.spyOn(api, 'getVouchers')
			.mockResolvedValue(fakeResponse);

		wrapper = mountWithProviders(
			<Provider store={realStore}>
				<Cart />
			</Provider>
		);

		const form = wrapper.find(Discount);
		await act(async () => {
			await flushPromises();

			form.find('input').simulate('change', { target: { value: code } });
			form.simulate('submit');
			wrapper.update();
		});

		expect(apiSpy).toHaveBeenCalled();
		expect(realStore.getState().cart.vouchers.length).toBe(1);
		expect(realStore.getState().cart.vouchers[0].code).toBe(code);
	});
});
