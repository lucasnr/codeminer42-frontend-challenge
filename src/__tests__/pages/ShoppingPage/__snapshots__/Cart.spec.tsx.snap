// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Cart> renders empty without fail and matches snapshot 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <ThemeProvider
    theme={
      Object {
        "background": Object {
          "primary": "#fdfdfd",
          "secondary": "#f6f6f6",
        },
        "colors": Object {
          "primary": "#266dd3",
          "secondary": "#344055",
        },
        "text": Object {
          "primary": "#333",
          "secondary": "#fff",
        },
        "title": "light",
      }
    }
  >
    <Cart>
      <styled.section>
        <section
          className="sc-fzoLsD lmTMkQ"
        >
          <styled.h2>
            <h2
              className="sc-fzpans VPZhh"
            >
              Shopping Cart
            </h2>
          </styled.h2>
          <EmptyCart>
            <styled.div>
              <div
                className="sc-AxheI ivOzZe"
              >
                <styled.h2>
                  <h2
                    className="sc-Axmtr lbvZms"
                  >
                    Looks like your cart is empty
                  </h2>
                </styled.h2>
                <styled.span>
                  <span
                    className="sc-AxmLO flfsQY"
                  >
                    :/
                  </span>
                </styled.span>
                <styled.h3>
                  <h3
                    className="sc-fzozJi jrzUig"
                  >
                    Try adding some products to your cart and they will show up here
                  </h3>
                </styled.h3>
              </div>
            </styled.div>
          </EmptyCart>
        </section>
      </styled.section>
    </Cart>
  </ThemeProvider>
</Provider>
`;

exports[`<Cart> renders with products without fail and matches snapshot 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <ThemeProvider
    theme={
      Object {
        "background": Object {
          "primary": "#fdfdfd",
          "secondary": "#f6f6f6",
        },
        "colors": Object {
          "primary": "#266dd3",
          "secondary": "#344055",
        },
        "text": Object {
          "primary": "#333",
          "secondary": "#fff",
        },
        "title": "light",
      }
    }
  >
    <Cart>
      <styled.section>
        <section
          className="sc-fzoLsD lmTMkQ"
        >
          <styled.h2>
            <h2
              className="sc-fzpans VPZhh"
            >
              Shopping Cart
            </h2>
          </styled.h2>
          <styled.div>
            <div
              className="sc-fzplWN grCRTV"
            >
              <ul>
                <Product
                  available={9}
                  id={5}
                  key="5"
                  name="Strawberry"
                  price={15}
                  quantity={1}
                >
                  <styled.li>
                    <li
                      className="sc-AxjAm jqfJtR"
                    >
                      <img
                        alt="Strawberry"
                        src="https://dumont.sp.gov.br/lib/img/no-image.jpg"
                      />
                      <styled.div>
                        <div
                          className="sc-AxiKw cLgGgt"
                        >
                          <styled.h4>
                            <h4
                              className="sc-AxirZ cvDiiJ"
                            >
                              Strawberry
                            </h4>
                          </styled.h4>
                          <styled.div>
                            <div
                              className="sc-AxhCb feUlwF"
                            >
                              <span>
                                Quantity: 
                                1
                              </span>
                              <span>
                                $ 
                                15.00
                              </span>
                            </div>
                          </styled.div>
                        </div>
                      </styled.div>
                      <styled.div>
                        <div
                          className="sc-AxhUy jJYQuF"
                        >
                          <button
                            aria-label="Increase amount on cart"
                            onClick={[Function]}
                          >
                            +
                          </button>
                          <button
                            aria-label="Decrease amount on cart"
                            onClick={[Function]}
                          >
                            -
                          </button>
                        </div>
                      </styled.div>
                    </li>
                  </styled.li>
                </Product>
              </ul>
              <Discount>
                <styled.form
                  onSubmit={[Function]}
                >
                  <form
                    className="sc-AxgMl ixeDyB"
                    onSubmit={[Function]}
                  >
                    <input
                      onChange={[Function]}
                      placeholder="Discount code"
                      type="text"
                      value=""
                    />
                    <button
                      type="submit"
                    >
                      Apply
                    </button>
                  </form>
                </styled.form>
              </Discount>
            </div>
          </styled.div>
          <styled.ul>
            <ul
              className="sc-fznyAO cTnyjm"
            >
              <li>
                <span>
                  Subtotal
                </span>
                <span
                  data-testid="subtotal"
                >
                  $ 
                  15.00
                </span>
              </li>
              <li>
                <span>
                  Shipping
                </span>
                <span
                  data-testid="shipping"
                >
                  $ 
                  30.00
                </span>
              </li>
              <styled.li>
                <li
                  className="sc-fznKkj cWqYpR"
                >
                  <span>
                    Discount
                  </span>
                  <span
                    data-testid="discount"
                  >
                    $ 
                    0.00
                  </span>
                </li>
              </styled.li>
              <li>
                <span>
                  Total
                </span>
                <span
                  data-testid="total"
                >
                  $ 
                  45.00
                </span>
              </li>
            </ul>
          </styled.ul>
        </section>
      </styled.section>
    </Cart>
  </ThemeProvider>
</Provider>
`;
